lib/Fmt_ast.ml

```diff
@@ -2691,16 +2683,46 @@ and fmt_expression c ?(box = true) ?pro ?epi ?eol ?parens ?(indent_wrap = 0)
                $ fmt_atrs ) ) )
   | Pexp_setinstvar (name, expr) ->
       hvbox 0
-        (Params.Exp.wrap c.conf c.source ~loc:pexp_loc ~parens
+        (Params.Exp.wrap c.conf ~parens
            ( Params.parens_if has_attr c.conf
                ( fmt_str_loc c name $ fmt_assign_arrow c
                $ hvbox 2 (fmt_expression c (sub_exp ~ctx expr)) )
            $ fmt_atrs ) )
   | Pexp_poly _ ->
       impossible "only used for methods, handled during method formatting"
   | Pexp_hole -> hvbox 0 (fmt_hole () $ fmt_atrs)
+  | Pexp_beginend e ->
+      let wrap_beginend =
+        match ctx0 with
+        (* begin-end keywords are handled when printing if-then-else
+           branch *)
+        | Exp {pexp_desc= Pexp_ifthenelse (_, _, Some z); _}
+          when Base.phys_equal xexp.ast z ->
+            Fn.id
```

Comment:
This is already better than before but is here any chance of implementing the special formatting here instead ? Perhaps `fmt_expression` could take a `?hint` argument that would optionally insert a break before every expression, except beginend, which would break after the keyword.

Reply:
I'm not sure it would make the code clearer, I like how we managed to put all the treatment in this Pexp_beginend.